File: enode/src/main/java/org/enodeframework/queue/command/DefaultCommandService.java
package org.enodeframework.queue.command;

import org.enodeframework.commanding.CommandResult;
import org.enodeframework.commanding.CommandReturnType;
import org.enodeframework.commanding.ICommand;
import org.enodeframework.commanding.ICommandService;
import org.enodeframework.common.serializing.ISerializeService;
import org.enodeframework.common.utilities.Ensure;
import org.enodeframework.common.utilities.RemotingUtil;
import org.enodeframework.queue.ISendMessageService;
import org.enodeframework.queue.QueueMessage;

import java.util.concurrent.CompletableFuture;

/**
 * @author anruence@gmail.com
 */
public class DefaultCommandService implements ICommandService {

    private final String topic;

    private final String tag;

    private final ISendMessageService sendMessageService;

    private final ICommandResultProcessor commandResultProcessor;

    private final ISerializeService serializeService;

    public DefaultCommandService(String topic, String tag, ICommandResultProcessor commandResultProcessor, ISendMessageService sendMessageService, ISerializeService serializeService) {
        this.topic = topic;
        this.tag = tag;
        this.commandResultProcessor = commandResultProcessor;
        this.sendMessageService = sendMessageService;
        this.serializeService = serializeService;
    }

    @Override
    public CompletableFuture<Void> sendAsync(ICommand command) {
        return sendMessageService.sendMessageAsync(buildCommandMessage(command, false));
    }

    @Override
    public CompletableFuture<CommandResult> executeAsync(ICommand command) {
        return executeAsync(command, CommandReturnType.CommandExecuted);
    }

    @Override
    public CompletableFuture<CommandResult> executeAsync(ICommand command, CommandReturnType commandReturnType) {
        CompletableFuture<CommandResult> taskCompletionSource = new CompletableFuture<>();
        try {
            Ensure.notNull(commandResultProcessor, "commandResultProcessor");
            commandResultProcessor.registerProcessingCommand(command, commandReturnType, taskCompletionSource);
            CompletableFuture<Void> sendMessageAsync = sendMessageService.sendMessageAsync(buildCommandMessage(command, true));
            sendMessageAsync.thenAccept(sendResult -> {
            }).exceptionally(ex -> {
                commandResultProcessor.processFailedSendingCommand(command);
                taskCompletionSource.completeExceptionally(ex);
                return null;
            });
        } catch (Exception ex) {
            taskCompletionSource.completeExceptionally(ex);
        }
        return taskCompletionSource;
    }

    protected QueueMessage buildCommandMessage(ICommand command, boolean needReply) {
        Ensure.notNull(command.getAggregateRootId(), "aggregateRootId");
        Ensure.notNull(topic, "topic");
        String commandData = serializeService.serialize(command);
        String replyAddress = needReply && commandResultProcessor != null ? RemotingUtil.parseAddress(commandResultProcessor.getBindAddress()) : null;
        CommandMessage commandMessage = new CommandMessage();
        commandMessage.setCommandData(commandData);
        commandMessage.setReplyAddress(replyAddress);
        commandMessage.setCommandType(command.getClass().getName());
        String messageData = serializeService.serialize(commandMessage);
        QueueMessage queueMessage = new QueueMessage();
        queueMessage.setTopic(topic);
        queueMessage.setTag(tag);
        queueMessage.setBody(messageData);
        queueMessage.setRouteKey(command.getAggregateRootId());
        String key = String.format("%s%s", command.getId(), command.getAggregateRootId() == null ? "" : "_cmd_agg_" + command.getAggregateRootId());
        queueMessage.setKey(key);
        return queueMessage;
    }
}

File: enode/src/main/java/org/enodeframework/queue/command/DefaultCommandService.kt
package org.enodeframework.queue.command

import org.enodeframework.commanding.CommandResult
import org.enodeframework.commanding.CommandReturnType
import org.enodeframework.commanding.ICommand
import org.enodeframework.commanding.ICommandService
import org.enodeframework.common.serializing.ISerializeService
import org.enodeframework.common.utils.Assert
import org.enodeframework.common.utils.ReplyUtil
import org.enodeframework.queue.ISendMessageService
import org.enodeframework.queue.QueueMessage
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * @author anruence@gmail.com
 */
class DefaultCommandService(
    private val topic: String,
    private val tag: String,
    private val commandResultProcessor: ICommandResultProcessor,
    private val sendMessageService: ISendMessageService,
    private val serializeService: ISerializeService,
) : ICommandService {
    override fun sendAsync(command: ICommand): CompletableFuture<Boolean> {
        return sendMessageService.sendMessageAsync(buildCommandMessage(command, false))
    }

    override fun send(command: ICommand): Boolean {
        return this.sendAsync(command).join()
    }

    override fun executeAsync(command: ICommand): CompletableFuture<CommandResult> {
        return executeAsync(command, CommandReturnType.CommandExecuted)
    }

    override fun executeAsync(
        command: ICommand,
        commandReturnType: CommandReturnType
    ): CompletableFuture<CommandResult> {
        val taskCompletionSource = CompletableFuture<CommandResult>()
        try {
            Assert.nonNull(commandResultProcessor, "commandResultProcessor")
            commandResultProcessor.registerProcessingCommand(command, commandReturnType, taskCompletionSource)
            val sendMessageAsync = sendMessageService.sendMessageAsync(buildCommandMessage(command, true))
            sendMessageAsync.exceptionally { ex: Throwable ->
                commandResultProcessor.processFailedSendingCommand(command)
                taskCompletionSource.completeExceptionally(ex)
                null
            }
        } catch (ex: Exception) {
            taskCompletionSource.completeExceptionally(ex)
        }
        return taskCompletionSource
    }

    override fun execute(command: ICommand): CommandResult {
        return this.executeAsync(command).join()
    }

    override fun execute(command: ICommand, commandReturnType: CommandReturnType): CommandResult {
        return this.executeAsync(command, commandReturnType).join()
    }

    protected fun buildCommandMessage(command: ICommand, needReply: Boolean): QueueMessage {
        Assert.nonNull(command.aggregateRootId, "aggregateRootId")
        Assert.nonNull(topic, "topic")
        val commandData = serializeService.serialize(command)
        val commandMessage = CommandMessage()
        if (needReply) {
            commandMessage.replyAddress = ReplyUtil.toUri(commandResultProcessor.bindAddress)
        }
        commandMessage.commandData = commandData
        commandMessage.commandType = command.javaClass.name
        val messageData = serializeService.serialize(commandMessage)
        val queueMessage = QueueMessage()
        queueMessage.topic = topic
        queueMessage.tag = tag
        queueMessage.body = messageData
        queueMessage.routeKey = command.aggregateRootId
        val key = String.format(
            "%s%s", command.id, Optional.ofNullable(command.aggregateRootId)
                .map { x: String -> "_cmd_agg_$x" }.orElse("")
        )
        queueMessage.key = key
        return queueMessage
    }
}